### ISSUE TITLE
Expose way to parse `AttributeArgs` from `proc_macro2::TokenStream`

### ISSUE THREAD

**Issue by casey** [dtolnay/syn#788]
TL;DR To summarize, I'd like to write an attribute proc_macro entry point, but with proc_macro2 types, for testability,  but I can't figure out a good way to to do `parse_macro_input!(attr as AttributeArgs)`, where `attr` is a `proc_macro2::TokenStream`.

First off, thank you very much for syn. I'm writing my first procedural macros, and it's very complex, but it's also clear that a whole lot of work has gone into making it easier, and into creating supporting infrastructure, which I greatly appreciate.

I might be having an X / Y problem. I'll try to be as descriptive as possible, in case there's actually a better solution to do what I want that I'm not seeing.

I'm trying to write a procedural macro which is an attribute on traits, so you an do:

```rust
#[my_crate::my_proc_macro]
trait Foo {
}
```

And it will generate an impl of that trait. (This is a little backwards from the usual way of deriving implementations of traits from structs, but it makes more sense for my problem domain for the user to describe a trait, and then have the macro generate an impl.)

I am using the [darling crate](https://github.com/TedDriggs/darling), which provides serde-like derive for attribute parsers. I got the initial scaffolding working like this:

```rust
use proc_macro::TokenStream;

#[derive(FromMeta)]
struct MyProcMacro {}

#[proc_macro_attribute]
pub fn my_proc_macro(attr: TokenStream, item: TokenStream) -> TokenStream {
  let attr = parse_macro_input!(attr as AttributeArgs);
  let item = parse_macro_input!(item as ItemTrait);
  let table = MyProcMacro::from_list(&attr).unwrap();
  // would normally do stuff here, but let's just return the unmodified item
  item.into_token_stream().into()
}
```

However, this presents problems for testability. Since the inputs and outputs of `my_proc_macro` are of type `proc_macro::TokenStream`, and not `proc_macro2::TokenStream`, I can't test this function directly.

So, I set out to write something that would have this signature:

```rust
fn my_proc_macro_inner(
  attr: proc_macro2::TokenStream,
  item: proc_macro2::TokenStream
) -> proc_macro2::TokenStream {
  // stuff
}
```

By taking and returning `proc_macro2::TokenStream`, I could test this function from normal unit tests.

However, I ran into an issue. There appears to be no public way to parse an `AttributeArgs` (which is just a ` Vec<NestedMeta>`) from a `proc_macro2::TokenStream`.

Just trying to change the types and invoking `parse_macro_input!()` fails with mismatched types:

```rust
fn parse_attr(attr: proc_macro2::TokenStream) -> proc_macro2::TokenStream {
  let args = parse_macro_input!(attr as AttributeArgs); // can't call this inside function with this signature
  unimplemented!()
}
```

The code that parses `AttributeArgs` is [relatively simple](https://docs.rs/syn/1.0.18/src/syn/parse_macro_input.rs.html#92), but it looks like it depends on a ParseStream, which I can't construct, so I can't write my own implementation.

Is there a better way to do what I want, or an alternative way to parse an `AttributeArgs`?

**Comment by casey** [comment]
Fixed some typos!

**Comment by dtolnay** [comment]
The closest today would be:

```rust
use syn::parse::Parser;
use syn::punctuated::Punctuated;
use syn::{NestedMeta, Token};

pub fn demo(input: proc_macro2::TokenStream) {
    let parser = Punctuated::<NestedMeta, Token![,]>::parse_terminated;
    let attribute_args = match parser.parse2(input) {
        Ok(args) => args,
        Err(err) => ...,
    };


}
```

This gives you `Punctuated<NestedMeta, Token![,]>` instead of `Vec<NestedMeta>` but should be usable for most of the same purposes.

**Comment by casey** [comment]
This is perfect, thank you!

**Comment by casey** [comment]
I think this can be closed, so doing so. Although what do you think about a hint in the docs for `AttributeArgs` suggesting what you wrote? It might be a nice breadcrumb for anyone doing the same. If you think that's a good idea, I'd be happy to open a PR.

**Comment by notdanilo** [comment]
A clear API for this is far more desired. But if this isn't a priority, the solution above should be added to the documentation. I had to dig the issues to find this.

**Comment by hecatia-elegua** [comment]
I will add onto this by saying that the docs, mixed with tutorials on proc macros are really confusing as they recommend different ways of doing these things.
One of the nice things about rust is that you generally have one way of doing things or clear differentation in docs.

[example-of-a-derive-macro](https://docs.rs/syn/1.0.107/syn/index.html#example-of-a-derive-macro) and [parse_macro_input](https://docs.rs/syn/1.0.107/syn/macro.parse_macro_input.html#intended-usage) and [fathomable-rust-macros](https://geeklaunch.io/blog/fathomable-rust-macros/#communicating-with-the-compiler) say to use `parse_macro_input!` in proc-macros.

[ferrous-systems](https://github.com/ferrous-systems/testing-proc-macros/blob/e431fcdc541bcee976976c1a5c0fc956944ec220/src/parse.rs#L12) says to first convert to `proc_macro2::TokenStream` and then use `syn::parse2`. I also remember some others recommending to first do this `.into()`-conversion.
Ah.
[proc_macro2](https://docs.rs/proc-macro2/latest/proc_macro2/#usage) says _if_ using `Syn`, one should use `parse_macro_input!`.

Then my questiong would be: is `parse_macro_input!` basically replacing ferrous-systems `parse` function?
